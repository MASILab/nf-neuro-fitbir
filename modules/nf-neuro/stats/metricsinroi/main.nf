process STATS_METRICSINROI {
    tag "$meta.id"
    label 'process_single'

    container "scilus/scilpy:2.2.0_cpu"

    input:
    tuple val(meta), path(metrics), path(rois), path(rois_lut)  /* optional, input = [] */

    output:
    tuple val(meta), path("*_stats.json")       , emit: mqc
    path "versions.yml"                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def suffix = task.ext.first_suffix ? "${task.ext.first_suffix}_stats" : "stats"
    def bin = task.ext.bin ? "--bin " : ""
    def normalize_weights = task.ext.normalize_weights ? "--normalize_weights " : ""
    def use_label = task.ext.use_label ? true : false
    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    if $use_label;
    then
        if [[ ! -f "$rois_lut" ]];
        then
            echo "ROI LUT is missing. Will fail."
        fi

        scil_volume_stats_in_labels $rois $rois_lut \
            --metrics $metrics \
            --sort_keys > ${prefix}__${suffix}.json
    else
        scil_volume_stats_in_ROI $rois \
            --metrics $metrics \
            --sort_keys \
            $bin $normalize_weights > ${prefix}__${suffix}.json
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def suffix = task.ext.first_suffix ? "${task.ext.first_suffix}_stats" : "stats"
    """
    scil_volume_stats_in_ROI -h
    scil_volume_stats_in_labels -h

    touch ${prefix}__${suffix}.json

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """
}
