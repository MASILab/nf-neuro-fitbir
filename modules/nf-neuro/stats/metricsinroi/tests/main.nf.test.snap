{
    "stats - metrics in roi -stub-run": {
        "content": [
            [
                "versions.yml:md5,03667d993a76f6643a67225c1099ea77"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-22T21:11:44.526236431"
    },
    "stats - metrics in ROI": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__stats.json:md5,7291461646d32ae71f7a248655783f32"
                    ]
                ],
                "1": [
                    "versions.yml:md5,03667d993a76f6643a67225c1099ea77"
                ],
                "mqc": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__stats.json:md5,7291461646d32ae71f7a248655783f32"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,03667d993a76f6643a67225c1099ea77"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-22T21:11:19.427514603"
    },
    "stats - metrics in Label": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__stats.json:md5,5bf37c68f8201323fe9d468cf271c4d8"
                    ]
                ],
                "1": [
                    "versions.yml:md5,03667d993a76f6643a67225c1099ea77"
                ],
                "mqc": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__stats.json:md5,5bf37c68f8201323fe9d468cf271c4d8"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,03667d993a76f6643a67225c1099ea77"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-09-22T21:11:31.707330877"
    }
}